레스토랑 조회 구현하기

solmatches opened the pull request:
안녕하세요! 5주차 레스토랑 조회 구현 과제 제출합니다:)

지역 목록 구현까지 작업 완료하였고
카테고리, 레스토랑 목록을 이어서 구현해나갈 예정입니다!

solmatches committed:
feat: 레스토랑 지역 목록 렌더 테스트 추가

solmatches committed:
feat: 레스토랑 지역 목록 화면 구현

solmatches committed:
feat: redux 적용

solmatches committed:
feat: redux mock 테스트 코드 추가

solmatches committed:
refactor: regions mock 데이터 fixture로 분리

solmatches committed:
refactor: 지역 목록을 Regions 컴포넌트로 분리

solmatches committed:
refactor: 지역 목록 redux 연동 부분 분리

RegionsContainer로 분리

solmatches committed:
feat: 지역 목록 api 연결 및 리팩토링

solmatches committed:
feat: 지역 목록을 가져오지 못한 경우 테스트 작성

solmatches committed:
refactor: 지역 목록을 가져오지 못한경우 에러메세지 구현

solmatches committed:
feat: reducer 테스트 작성

solmatches committed:
feat: errorMesage 파일 내부 baseErrorMessage 테스트 작성

hannut91 commented on src/components/Regions.jsx:
```diff
@@ -0,0 +1,19 @@
+import ERROR_MESSAGES from '../../constants/errorMessages';
+
+export default function Regions({ regions }) {
+  if (!regions) {
+    return <div>{ERROR_MESSAGES.REGION}</div>;
+  }
```
처음 앱에 집입했을 때, 지역목록이 없어서 안보이는 경우에도 에러라고 뜰 것 같네요. 메시지의 이름을 수정하던지, 아니면 에러일 경우를 따로 구분하는게 좋을 것 같아요

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,17 @@
+
+    const state = reducer(initialState, setRegions(regions));
+
+    expect(state.regions).toHaveLength(2);
```
왜 숫자 2 인지 이해하기 어려운 것 같아요. `regions`를 가지고 표현해주는 것은 어떨까요?

hannut91 commented on src/components/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,25 @@
+    regions,
+  }));
+
+  test('지역 목록을 가져오는 함수를 호출한다.', () => {
```
이미 위에서 `describe`를 사용하고 있으니 `it`를 사용하는게 좋을 것 같습니다.

hannut91 commented on src/components/Regions.test.jsx:
```diff
@@ -0,0 +1,25 @@
+import Regions from './Regions';
+
+describe('Regions', () => {
+  context('지역 목록을 가져왔으면', () => {
```
지역 목록을 API호출을 통해서 가져온 정보라서, `가져왔으면`이라고 작성하신 것 같아요.

이 함수의 입장에서는 API를 통해서 가져오는지 모르므로, API에 대한 컨텍스트가 이 컴포넌트 컨텍스트로 넘친 것 같아요. 컨텍스트가 서로 침범하면 변경하기 어려워 집니다. 예를들어서 지역 목록을 가져오지 못할 경우 기본 목록을 보여주는 기능을 추가한다고 해봅시다. 그러면 이 지역 목록을 가져오는 것이 아니라 원래 있는건데 헷갈릴 수 있겠죠.

solmatches committed:
feat: actions loadRegions 테스트 작성

solmatches committed:
feat: actions loadRegions 테스트 통과 구현

solmatches committed:
refactor: errorMessages -> emptyMessages로 변경

지역목록이 없을 때도 지역목록을 가져오지 못했다는 에러 메세지가 출력되는 문제
- 에러 메세지는 별도로 처리하고 목록이 없을 때 나타나는 문구로 변경

solmatches committed:
refactor: setRegion reducer의 테스트toHaveLength 값 개선

state.regions 의 개수가 왜 2인지 코드만으로 이해할 수 없어 `regions.length`로 대체

solmatches committed:
refactor: 테스트 코드의 가독성을 위해 test -> it 으로 변경

solmatches committed:
refactor: Regions 컴포넌트의 역할에 맞게 테스트 문구 수정

이 함수는 api를 통해서 지역 목록을 가져오는지 모른다.
컨텍스트가 서로 침범하면 변경하기 어려워지는 것을 명심

solmatches committed:
fix: store 경로 수정

solmatches committed:
refactor: 지역목록 조회 실패시 에러메세지 구현 추가

solmatches committed:
feat: reducer setErrorMessage 테스트 추가

solmatches committed:
feat: api 테스트 작성

solmatches committed:
feat: action이 없을 때 reducer 테스트 작성

solmatches committed:
refactor: reducer action에 기본값 추가

solmatches committed:
feat: reducer state가 undefined 일 때 테스트 추가

solmatches committed:
fix: reducer 반환값 수정

solmatches committed:
feat: reducer setErrorMessage가 없을 때 테스트 추가

solmatches committed:
refactor: errorMessage가 없을 때 setErrorMessage 구현

solmatches committed:
feat: CategoriesContainer 테스트 작성

solmatches committed:
feat: CategoriesContainer 구현

hannut91 commented on constants/emptyMessages.test.js:
```diff
@@ -0,0 +1,20 @@
+
+describe('emptyMessages', () => {
+  context('전달된 `text` 인자가 없으면', () => {
+    test('"목록"이 비어있다는 메세지를 반환한다', () => {
```
Describe-Context-It을 쓴다면 `it`를 사용하는게 좋을 것 같아요

hannut91 commented on src/components/Regions.jsx:
```diff
@@ -0,0 +1,19 @@
+
+export default function Regions({ regions, errorMessage }) {
+  if (!regions) {
+    return <div>{errorMessage || ERROR_MESSAGES.REGION}</div>;
```
```js
    return <div>{errorMessage ?? ERROR_MESSAGES.REGION}</div>;
```

`??`을 사용하는게 더 좋을 것 같아요. 왜 그런지는 아래 문서를 읽어봅시다

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

solmatches committed:
feat: Categories 테스트 작성

solmatches committed:
fix: 지역 목록 에러메세지 반환 연산자 개선

`||` or 연산자의 경우 falsy한 값인 `0`, `''` 값을 사용하려는 경우에도 기본값이 사용되므로
nullish한 값(null, undefined)일 때만 기본 값이 사용되도록 ?? 연산자를 사용하도록 변경함

solmatches committed:
fix: describe-context-test 구조일 때 가독성을위해 it 으로 변경

solmatches committed:
feat: Categories 컴포넌트 구현

solmatches committed:
refactor: App 테스트 코드 수정

solmatches committed:
refactor: categories mock data 추가 및 적용

solmatches committed:
fix: Regions 컴포넌트 에러메세지 노출 식 수정

regions 의 기본값이 빈배열이라 없는값으로 여기지 않아 에러 발생시에도 에러가 노출되지 않음

solmatches committed:
refactor: Categories store 데이터 연결

solmatches committed:
fix: setCategories actions 오타 수정

solmatches committed:
feat: fetchCategories action 테스트 작성

solmatches committed:
refactor: fetchCategories 구현

solmatches committed:
feat: fetchCategories api test 작성

solmatches committed:
feat: categories 관련 action, reducer 테스트 작성

solmatches committed:
feat: Categories 컴포넌트 버튼 V표시 추가 테스트 작성

solmatches committed:
refactor: Categories 버튼 V표시 테스트 통과 구현

solmatches committed:
refactor: Categories 컴포넌트에 errorMessage props 전달

solmatches committed:
feat: Categories 컴포넌트 클릭 이벤트 테스트 작성

solmatches committed:
refactor: Categories 컴포넌트 클릭시 V 표시 구현

solmatches committed:
feat: CategoriesContainer click 이벤트 작성

solmatches committed:
feat: CategoriesContainer click 이벤트 성공 구현

solmatches committed:
feat: reducer selectedCategoryId 테스트 추가

solmatches committed:
feat: reducer selectedCategoryId 클릭 이벤트 연동 구현

solmatches committed:
feat: Region 컴포넌트 클릭 테스트 작성

solmatches committed:
feat: Region 컴포넌트 클릭이벤트 구현

solmatches committed:
feat: RegionsContainer click 이벤트 테스트 추가

solmatches committed:
feat: RegionsContainer click 이벤트 성공 구현

solmatches committed:
feat: reducer selectRegionId 액션 테스트 작성

solmatches committed:
feat: reducer selectedRegionId 값 추가

solmatches committed:
feat: RestaurantsContainer 테스트 작성

solmatches committed:
feat: RestaurantsContainer 성공 구현

solmatches committed:
feat: fetchRestaurant api 호출 테스트 작성

solmatches committed:
feat: fetchRestaurants api 성공 구현

hannut91 commented on src/App.jsx:
```diff
@@ -1,5 +1,11 @@
+import CategoriesContainer from './components/CategoriesContainer';
+import RegionsContainer from './components/RegionsContainer';
+
 export default function App() {
   return (
-    <></>
+    <>
+      <RegionsContainer />
+      <CategoriesContainer />
+    </>
   );
```
임포트는 CategoriesContainer부터 하는데, 실제로 쓰이는 것은 `RegionsContainer`부터 사용되는 것 같네요. 순서를 맞추면 좋을 것 같습니다

hannut91 commented on src/components/Categories.jsx:
```diff
@@ -0,0 +1,24 @@
+      <ul>
+        {categories.map(({ id, name }) => (
+          <li key={id}>
+            <button type="button" onClick={() => onClick(id)}>
+              {name}
+              {selectedId === id && '(V)'}
+            </button>
```
만약에 선택됨을 표시할 때, `(V)`로 하는 것이 아니라 색깔을 다르게 표시한다던지 하는 방식을 변경할 경우 이 `<Categories />`컴포넌트를 수정해야 합니다. 

이러한 변경사항이 생겨도, 이러한 처리를 담당하는 버튼 컴포넌트로 별도의 컴포넌트로 분리해 봅시다.

hannut91 commented on src/components/Categories.test.jsx:
```diff
@@ -0,0 +1,35 @@
+
+const onClick = jest.fn();
+
+describe('Categories', () => {
+  it('화면에 카테고리 값이 나온다', () => {
+    const { getByText } = render(<Categories categories={categories} onClick={onClick} />);
+
+    const categoryNameButton = getByText(categories[0].name, { exact: false });
+
+    expect(categoryNameButton).toBeInTheDocument();
+  });
+
+  it('카테고리를 클릭하면 해당 카테고리에 V 표시가 붙는다', () => {
+    const { getByText } = render(
+      <Categories
+        categories={categories}
+        selectedId={categories[0].id}
+        onClick={onClick}
+      />,
+    );
+
+    const categoryNameButton = getByText(categories[0].name, { exact: false });
+
+    fireEvent.click(categoryNameButton);
+
+    expect(onClick).toBeCalledWith(categories[0].id);
+
+    expect(categoryNameButton).toHaveTextContent(/V/g);
+  });
+});
```
테스트를 보면 카테고리 목록을 표시하는 일과, 선택했을 때 선택됨을 표시하는 일 두 가지를 테스트하고 있죠.

테스트할 떄 뭔가 여러가지 일을 테스트하고 있는 것을 발견한다면, 분리해야 할 신호로 받아들이시면 좋습니다

hannut91 commented on src/components/Regions.jsx:
```diff
@@ -0,0 +1,24 @@
+      <ul>
+        {regions.map(({ id, name }) => (
+          <li key={id}>
+            <button type="button" onClick={() => onClick(id)}>
+              {name}
+              {selectedId === id && '(V)'}
+            </button>
```
`<Categories />`에서 하는 일과 똑같은 일을 여기서도 하고 있는 것 같아요.

hannut91 commented on src/services/api.test.js:
```diff
@@ -0,0 +1,41 @@
+  });
+
+  test('fetchRestaurants', async () => {
+    fetch.mockImplementationOnce(() => ({ json: () => Promise.resolve(mockRestaurants) }));
```
```js
    fetch.mockImplementationOnce(() => ({ json: async () => mockRestaurants }));
```

아마 이렇게 해도 될 것 같아요. 왜냐하먄 async함수는 반환타입이 프로미스면 그냥 프로미스를 반환하고, 일반값이라면 프로미스 객체로 감싼 객체를 반환하기 때문이에요

solmatches committed:
feat: loadRestaurants actions 테스트 작성

solmatches committed:
fix: selectedRegionId -> selectedRegion 변경

레스토랑 호출시 전달할 파라미터로 지역 이름이 필요하여 변경

solmatches committed:
feat: setRestaurant reducer 테스트 작성

solmatches committed:
feat: setRestaurants reducer 구현

solmatches committed:
refactor: RestaurantContainer 선택한 지역, 카테고리 없을 때 테스트 추가

solmatches committed:
feat: Restaurant 테스트 작성

solmatches committed:
feat: Restaurants 구현

solmatches committed:
fix: Restaurants 테스트 수정

hannut91 commented on src/redux/reducer.test.js:
```diff
@@ -0,0 +1,128 @@
+
+  describe('state가 undefined면', () => {
+    it('initialState를 반환한다', () => {
+      const initialState = {
+        regions: [],
+        errorMessage: null,
+      };
```
이 데이터가 반복적으로 사용되고 있는 것 같아요. 위에 선언해서 중복을 제거하면 좋을 것 같습니다

solmatches committed:
fix: selectedCategoryId falsy 예외처리 수정

id값이 0 인 경우에도 false로 처리되는 이슈

solmatches committed:
feat: App 레스토랑 목록 호출 test 추가

solmatches committed:
refactor: ErrorMessage 항목별로 구분하기

solmatches committed:
refactor: initialState 공통 변수로 분리 적용

solmatches committed:
refactor: 실제로 사용되는 컴포넌트 순서와 import 하는 순서 맞춤

solmatches committed:
feat: Categories에 categories 값이 없을 경우 테스트 추가

solmatches committed:
feat: Button 컴포넌트 테스트 작성

solmatches committed:
refactor: Button 테스트 name props -> children 변경

solmatches committed:
feat: Button 컴포넌트 구현

solmatches committed:
refactor: Regions에 Button 컴포넌트 적용

solmatches committed:
refactor: 선택한 버튼에 'V' 표시를 확인하는 테스트 제거

Categories, Regions 컴포넌트에 Button 컴포넌트를 적용하여
선택한 버튼에 대한 스타일링을 Button 컴포넌트에 위임하였으므로
해당 컴포넌트들의 역할에서 벗어남.

solmatches committed:
refactor: promise 반환 mock 함수 형식을 async로 간결하게 변경

async 함수는 반환값의 타입이 Promise면 Promise를 반환하고
Promise가 아니면 Promise 객체로 감싼 객체를 반환한다.

solmatches committed:
refactor: loadRestaurants에 redux-thunk getState 적용

solmatches committed:
fix: loadRestaurants test에 getState mock 추가

solmatches committed:
refactor: get utils 추가 및 CategoryContainer에 적용

solmatches committed:
refactor: Restaurants, Regions에 get utils 적용

solmatches commented:
우연히 발견하여 여쭤봅니다. 주간회고를 일요일까지 작성 마쳤는데 이 PR이 머지 되지 못한 이유가 있을까요? 🥲 
(11시 57분에 올려서 그런가요..)
